File: C++/Letter Combinations of Phone Num/phoneNumbers.cpp
using namespace std;

class Solution
{
public:
    void phoneBFS(int index, vector<string> path)
    {
        if (path.size() == digits.length())
        {
            // path length is same as input
            string pathStr = "";
            for (int i = 0; i < path.size(); i++)
            {
                pathStr += path[i]
            }
            combinations.push_back(pathStr);
            return
        }

        // retrieving the possible combinations from the map
        string letters = phoneNums[digits[index]];

        for (int j = 0; j < letters.length(); j++)
        {
            path.push_back(letters[j])

                // getting next path
                phoneBFS(index + 1, path);

            // popping off last element in path for next iteration
            path.pop_back()
        }
    }
    vector<string> letterCombinations(string digits)
    {

        vector<string> combinations;

        if (digits.length() == 0)
        {
            return combinations;
        }

        unordered_map<string, string> phoneNums;

        phoneNums["2"] = "abc";
        phoneNums["3"] = "def";
        phoneNums["4"] = "ghi";
        phoneNums["5"] = "jkl";
        phoneNums["6"] = "mno";
        phoneNums["7"] = "pqrs";
        phoneNums["8"] = "tuv";
        phoneNums["9"] = "wxyz";

        // calling phoneBFS
        phoneBFS(0, []);

        return combinations;
    }
};

File: Python/Letter Combinations of Phone Num/phoneNumbers.py
class Solution:

    def letterCombinations(self, digits: str) -> List[str]:

        if len(digits) == 0:
            #empty input 
            return []

        numMap = {}
        numMap['2'] = 'abc'
        numMap['3'] = 'def'
        numMap['4'] = 'ghi'
        numMap['5'] = 'jkl'
        numMap['6'] = 'mno'
        numMap['7'] = 'pqrs'
        numMap['8'] = 'tuv'
        numMap['9'] = 'wxyz'
        
        def phoneBFS(index:int, path:List[str]):
            if len(path) == len(digits):
                #the path is as long as the input 
                #so we have to append the path to the main combinatiuons list and return 
                combinations.append("".join(path))
                return 
            #getting the letters from the dictionary 
            letters = numMap[digits[index]]

            for letter in letters:
                #appending path with possible letter
                path.append(letter)

                #getting the next letter 
                phoneBFS(index + 1, path)

                #removing the last letter of path to backtrack
                path.pop()

        
        combinations = []

        #calling recursive function 
        phoneBFS(0, [])

        return combinations